### PR TITLE
Avoid parse/unparse public ECC keys in PK with USE_PSA when !ECP_C

### LINKED ISSUE #7073: Avoid parse/unparse public ECC keys in PK with USE_PSA
Background: see #6009 and in particular https://github.com/Mbed-TLS/mbedtls/issues/6009#issuecomment-1367896727.

When `USE_PSA_CRYPTO` is enabled, the PK module should:
- at parsing time, store public ECC keys as a byte array (suitable as input to `psa_import_key()`) _in addition_ to storing them as an `ecp_keypair`
- when using the key (`pk_verify()`) or writing it out (`pk_write_pubkey*()`) use that array in preference to the `ecp_keypair` structure.

This reduces the reliance on ECP (and ultimately) bignum functions. Note: for now we're keeping the `ecp_keypair` structure for compatibility reasons, due to the existence of the public API function `mbedtls_pk_ec()`. In the future we'll probably remove it when `ECP_C` is disabled.

Suggested course of action:
1. Add two new members (pointer, length) to `mbedtls_pk_context` to hold the serialized key. Adapt `pk_init()` and `pk_free()` accordingly.
2. Parsing: all possible ECC public key formats end up calling `pk_get_ecpubkey()`, which takes as input exactly the EC Point that `psa_import_key()` wants. Make this function copy its input to the new members of the `mbedtls_pk_context`. (Note: actually this only works for uncompressed points (ie, input starts with 0x04). For compressed points, until PSA supports them, we have no choice but to call `mbedtls_ecp_write_point_binary()` to get the uncompressed form from `key->Q`. This leave one parse-serialize cycle, but at least it's isolated in `pk_get_ecpubkey()`.
3. Verify: in `ecdsa_verify_wrap()` we no longer need to call `mbedtls_ecp_point_write_binary()` and can directly get the input to `psa_import_key()` from the new members of `pk_context`.
4. Write: similarly to parse, all possible formats end up calling `pk_write_ec_pubkey()`. Again, instead of calling `mbedtls_ecp_point_write_binary()` we can just copy from the new members of `pk_context`.

Depends on: #6389 

Follow-up: same for private keys.

### PR DESCRIPTION
**Pull request by valeriosetti**
As suggested [here](https://github.com/Mbed-TLS/mbedtls/pull/7202#issuecomment-1508063453), this PR is the reshaped version of #7202 which includes only the changes which are relevant to solve issue #7073.

Depends on:
- ~#6838~
- #7595

Resolves #7073

## PR checklist

- [x] **changelog** not required - internal refactoring
- [x] **backport** just the bug fix in `test_suite_debug` https://github.com/Mbed-TLS/mbedtls/pull/7642
- [x] **tests** covered by existing tests

### CODE REVIEW

**Comment by valeriosetti** [comment]
Ok, even though I didn't list #7514 in the dependency list, it seems that its merging caused some conflict. I'll do a rebase

File: library/pk_wrap.c
Comment (mpg) [inline library/pk_wrap.c:740]: I'm not sure we need to copy to `curve` and `curve_bits` here, we could use the fields directly (after just checking that they're not 0).

File: library/pk_wrap.c
Comment (mpg) [inline library/pk_wrap.c:1113]: Can't we just use the new field `pk_ec_family` here?

File: include/mbedtls/pk.h
Comment (mpg) [reply include/mbedtls/pk.h:250]: Also: are those fields set for all PK contexts holding an ECC key (private or public), or only for those that hold a public key? Is the answer the same for all 4 of the new fields, or different for the last two?

File: library/pk.c
Comment (mpg) [inline library/pk.c:860]: This is an internal function, so I don't think these checks are needed.

File: library/pk.c
Comment (mpg) [inline library/pk.c:908]: Note: it doesn't look like this function is currently used anywhere (from grepping the source). On one hand, perhaps it will be useful later, but OTOH, not used means untested... Wdyt?

File: library/pk.c
Comment (mpg) [inline library/pk.c:883]: The comment should be updated to match the check, along the lines of "Only the following key types are supported."

(Same in the other function if we keep it.)

**Review by mpg — CHANGES_REQUESTED** [review]
This is clearly going in the right direction. I left a number of comments, but the general structure is looking good.

**Comment by mpg** [comment]
@valeriosetti Btw, can you also address the comments Jerry left [on 6838](https://github.com/Mbed-TLS/mbedtls/pull/6838#pullrequestreview-1414307538) as this is the first PR that builds on it?

File: include/mbedtls/pk.h
Comment (valeriosetti) [reply include/mbedtls/pk.h:251]: As for the `pub_raw` renaming, I just found out that actually `mbedtls_x509_buf pk_raw;` is a field of the `struct mbedtls_x509_crt` so it's even more suggested to rename this field to `pub_raw` to avoid confusion!

**Comment by valeriosetti** [comment]
> @valeriosetti Btw, can you also address the comments Jerry left [on 6838](https://github.com/Mbed-TLS/mbedtls/pull/6838#pullrequestreview-1414307538) as this is the first PR that builds on it?

Done! I just skipped [this comment](https://github.com/Mbed-TLS/mbedtls/pull/6838#discussion_r1185787155) because it was related to a line being too long, but due to the changes done in this PR it should be fine now.

File: tests/suites/test_suite_pk.function
Comment (mpg) [inline tests/suites/test_suite_pk.function:732]: This (and the next 3 similar additions) is a compatibility smell. What our test code is doing here would be perfectly legal (according to our documentation) for an application to do. Applications that were doing that would then be broken after this PR, which is not acceptable.

This PR is already fairly large as it is, so I've created a follow-up issue which I suggest you address right after this PR and #7438 have been merged: #7570

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function:732]: Based on the discussion in https://github.com/Mbed-TLS/mbedtls/issues/7570, I think this is not just a compatibility smell, but an outright design blocker. As soon as `mbedtls_pk_ec` is called on the PK context — and [we can't detect that](https://github.com/Mbed-TLS/mbedtls/issues/7570#issuecomment-1542186552) — the `priv_id` field becomes unusable. So the `priv_id` field is effectively unusable unless the `mbedtls_pk_ec` function doesn't exist. Since this function has to exist in all the cases where it exists in Mbed TLS 3.4.0 for backward compatibility, the thing we can do is:

* Make `mbedtls_pk_ec` not part of the public API if `MBEDTLS_ECP_LIGHT` is defined but not `MBEDTLS_ECP_C`. It can still be defined in an internal header, possibly returning a const pointer because we want it to be read-only access.
* Change `priv_id` and all the code that accesses it to be invoked only when `!defined(MBEDTLS_ECP_C)`.

Comment (valeriosetti) [reply tests/suites/test_suite_pk.function:732]: >  possibly returning a const pointer because we want it to be read-only access

Some tests use `mbedtls_pk_ec` to get access to the `ecp_keypair` directly to write it, just like the one that @mpg highlighted in this comment. I think that making `mbedtls_pk_ec` private is already enough (we have control on what the library's code does, but not the final user's one) since it allows to keep this test. Wdyt?

**Comment by valeriosetti** [comment]
Following the discussion in #7570 and suggestions there ([1](https://github.com/Mbed-TLS/mbedtls/issues/7570#issuecomment-1543599479) and [2](https://github.com/Mbed-TLS/mbedtls/issues/7570#issuecomment-1543620882)) I added the `need-work` label to this PR

**Comment by valeriosetti** [comment]
@mpg, @gilles-peskine-arm
Last 3 commits (https://github.com/Mbed-TLS/mbedtls/pull/7554/commits/779bf0fabfe261c4e7a9f2c6a325a9656847da21, https://github.com/Mbed-TLS/mbedtls/pull/7554/commits/84b357f1af2c0353368ea1da1fd58c30ee844e0a, https://github.com/Mbed-TLS/mbedtls/pull/7554/commits/0641ec90a796dce09758455ce04cdbd8ae1c64cd) try to address the changes requested/discussed in #7570. Can you please have a look and let me know if I got close to what you proposed?

File: tests/suites/test_suite_pk.function
Comment (mpg) [reply tests/suites/test_suite_pk.function]: See https://github.com/Mbed-TLS/mbedtls/issues/7570#issuecomment-1545443804 - I'm trying to regroup all discussion in one place as I find it harder to follow otherwise.

File: library/pk_internal.h
Comment (mpg) [inline library/pk_internal.h:64]: I don't like having two function with the same name but different signatures and contracts. I'd prefer having three functions for clarity:
1. Public `mbedtls_pk_ec()` - formally deprecated, hence never used by the library or tests (it is absent when `MBEDTLS_DEPRECATED_REMOVED` is set).
2. Internal function `mbedtls_pk_ec_ro()` returning a `const` pointer.
3. Internal function `mbedtls_pk_ec_rw()` returning a non-`const` pointer.

**Review by mpg — CHANGES_REQUESTED** [review]
Just commenting on the handling of `pk_ec()`.

File: library/pk_internal.h
Comment (mpg) [reply library/pk_internal.h:64]: Should have updated my comment, because we followed exactly the same path in private discussion with Valerio:
- it's to early to formally deprecate `pk_ec`, we must clarify replacements first;
- this should be its own PR indeed.

**Comment by valeriosetti** [comment]
Due to recent discussions in #7570 this PR already underwent a good amount of changes and some strategy's change as well. Since yesterday [we reached an agreement](https://github.com/Mbed-TLS/mbedtls/issues/7570#issuecomment-1547653771) on the final goal, I thought it might be helpful for a reviewer to have a clean history to analyze so I rebased on top of #7595

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:2494]: I'm confused. Don't we already have `component_test_psa_crypto_config_accel_all_ec_algs_use_psa` for that?

This PR is not just refactoring existing code in existing configs, so it shouldn't need new tests. (It's preparation work for making new configs possible in the future, namely enabling PK in `config_psa_crypto_full_all_ec_algs_no_ecp_use_psa`, but that will come later one 7074 is also done, and perhaps a small follow-up after that.)

Comment (valeriosetti) [reply tests/scripts/all.sh:2494]: This problem was raised by a CI failure with `test_psa_crypto_config_accel_all_curves_except_[p192/x25519]`. The error was in `pk_wrap.c`: it was picking the "wrong" the `ecdsa_verify_wrap`, the one using `mbedtls_ecdsa_read_signature`. This happened because, following our discussion(1), I changed the guards for that function from `USE_PSA_CRYPTO` to `MBEDTLS_PK_USE_PSA_EC_PUB_KEY`. The latter is only available when `!ECP_C`.

A part from having the CI happy, to me this change seems reasonable according to the fact that `ECP_LIGHT` was only introduced after these failing test cases. However, as far as they are concerned, we don't want the full `ECP_C` here, but only `ECP_LIGHT`. Wdyt?

(1) i.e. when `ECP_C` is defined we have to use the legacy ecp_keypair structure and legacy functions..

Comment (mpg) [reply tests/scripts/all.sh:2494]: > (1) i.e. when `ECP_C` is defined we have to use the legacy ecp_keypair structure and legacy functions..

I think there's a misunderstanding: when `USE_PSA` crypto is defined, we have to use PSA for crypto operations (except when that is not possible, which currently is only restartable operations). So, the decision tree is:
- if `!USE_PSA`, use legacy context and legacy functions (unchanged)
- if `USE_PSA`, use PSA functions, but:
  - if `ECP_C`: keep the key material in `ecp_keypair` and convert to PSA before each operation (also unchanged, except perhaps [re-organizing the code in order to share code with the next case](https://github.com/Mbed-TLS/mbedtls/issues/7570#issuecomment-1545520593))
  - if `!ECP_C`: we can keep the key material in new PSA-friendly fields to avoid conversions (that's new).

Again, this is a refactoring PR, configurations that used to work are supposed to keep working the same way.

Note that in [the code structure I suggested](https://github.com/Mbed-TLS/mbedtls/issues/7570#issuecomment-1545498889), in the `USE_PSA` case we are using PSA all the time and the only thing that varies depending on config (`ECP_C` or not) and key type (opaque or not, public or private) is how we transform (or not) the key material from the form we have to the form used by the internal function - but that function always uses PSA.

Comment (valeriosetti) [reply tests/scripts/all.sh:2494]: Ok, so I think I'll have to restructure the commits because I'm not sure I correctly addressed this point:

> if USE_PSA, use PSA functions, but:
    if ECP_C: keep the key material in ecp_keypair and convert to PSA before each operation (also unchanged, except perhaps https://github.com/Mbed-TLS/mbedtls/issues/7570#issuecomment-1545520593)

I'll try to do it asap

Comment (valeriosetti) [reply tests/scripts/all.sh]: @mpg I think I reshaped the code as you suggested and indeed the issue causing this problem/comment is not present any more. Since the CI is also happy, I think this PR is ready for a review ;)

**Review by mpg — CHANGES_REQUESTED** [review]
Looking pretty good to me! I could only find minor issues in comments :)

File: include/mbedtls/pk.h
Comment (valeriosetti) [reply include/mbedtls/pk.h:263]: Yes this is definitely a left over from the past. The idea is of course to update this comment in #7074 ;)

File: library/debug.c
Comment (valeriosetti) [reply library/debug.c:223]: Yes, I noticed that there was the `mbedtls_debug_print_buf()` which seemed to do what I was looking for, however that would have required a change in the test case because the output is not 100% identical.

> 2d EPIC when we try to make things work without BIGNUM_C

Oh wow, thanks to this comment I just realized that there is an extra list of EPICs that I was not aware of ':)

File: library/pkparse.c
Comment (mprse) [inline library/pkparse.c:725]:
```suggestion
            return MBEDTLS_ERR_PK_BUFFER_TOO_SMALL;
```

File: include/mbedtls/pk.h
Comment (mprse) [inline include/mbedtls/pk.h:285]:
```suggestion
     * ecp_keypair structure is used for storing the public key and performing
```

File: library/pk_wrap.c
Comment (mprse) [inline library/pk_wrap.c:1149]: I think we should check here if the conversion was successful and if not then return an error.

File: tests/suites/test_suite_pk.function
Comment (mprse) [inline tests/suites/test_suite_pk.function:36]: I think we should check here if the conversion was successful and if not then return an error.

**Review by mprse — COMMENTED** [review]
Changes looks good.
Left few comments and propositions for minor improvements.

File: library/debug.c
Comment (mpg) [reply library/debug.c:223]: Yeah, it's easy to miss as it's on another board (we have board for active EPIC and another one for the future), but I added a link from the first card in the 2c EPIC.

Currently it's not really populated (just 1 issue that you already resolved some time ago), and with a bit of luck it will turn out to be trivial, considering the way we're handling things in the 2c EPIC. Once we have PK working without `ECP_LIGHT` (which should be fairly soon now: this PR, its private counterpart, then probably just 1 more PR I hope), I'll re-assess the situation and see what dependencies on `BIGNUM_C` remain. I'm hoping at that point it will be just `debug.c`, which as we were saying will probably be quite easy to handle.

**Review by mpg — APPROVED** [review]
LGTM

**Comment by mpg** [comment]
@valeriosetti I think now would be a good time to prepare the partial backport (just the `test_suite_debug` fix), so we can merge this quickly once @mprse has re-reviewed.

**Comment by valeriosetti** [comment]
> @valeriosetti I think now would be a good time to prepare the partial backport (just the `test_suite_debug` fix), so we can merge this quickly once @mprse has re-reviewed.

Done! Please see https://github.com/Mbed-TLS/mbedtls/pull/7642

**Pull request merged**
